torch/_dynamo/guards.py

```diff
@@ -737,45 +776,90 @@ def compile_check_fn(
             if isinstance(guard, DuplicateInputs):
                 source_a = guard.input_source_a
                 source_b = guard.input_source_b
-                code_part = f"{source_a.name()} is {source_b.name()}"
+                code_part = CodePart(
+                    [source_a, source_b],
+                    f"{source_a.name()} is {source_b.name()}",
+                    "DuplicateInputs",
+                )
                 code_parts.append(code_part)
-                verbose_code_parts.append(code_part)
             else:
                 raise RuntimeError(f"Unknown GuardEnvExpr: {guard}")
 
         code_parts.extend(local_builder.shape_env_code)
-        verbose_code_parts.extend(local_builder.shape_env_code)
         assert not global_builder.shape_env_code
 
-        code = " and ".join(unique(code_parts))
         closure_vars = collections.OrderedDict(
             [
                 ("___guarded_code", self),
                 ("___check_tensors", check_tensors_fn),
-                ("___check_tensors_verbose", check_tensors_verbose_fn),
+                ("___check_tensors_verbose", check_tensor_verbose),
                 ("tensor_check_names", tensor_check_names),
+                ("part_map", part_map),
             ]
             + list(SYMPY_INTERP.items())
         )
         closure_vars.update(CLOSURE_VARS)
+
+        # Let's go over how this code works.
+        #
+        # 1) The cache entry structure in eval_frame.c is stored in the frame's extra field. Each cache_entry is a node
+        # in a linked list. Each cache entry represents a compiled frame with specializations. In order to know if
+        # a cache entry's compiled frame is valid for reuse, we need to invoke a function, check_fn,
+        # to compare current state against the specializations captured at compile time.
+        #
+        # 2) The function, check_fn, mentioned above, is defined by executing the function below, ___make_guard_fn
+        #
+        # 3) In this, this code is rather confusing, because it defines both ___make_guard_fn and the lambda it produces
+        # which becomes the check_fn.
```

Comment:
this... doesn't seem confusing to me? We need to create a closure for the lambda no? Is the confusion why we didn't just setup locals on the `eval` context directly so `__make_guard_fn` is not necessary?

Reply:
Well, I thought it was confusing the first time I saw it. Anything going into exec to make a new fn is a little confusing...